mrp_bom_attribute_match_semifinished_products/wizard/finished_product_structure_wizard.py

```diff
@@ -0,0 +1,136 @@
+from odoo import _, api, fields, models
+
+
+class FinishedProductStructureWizard(models.TransientModel):
+    _name = "finished.product.structure.wizard"
+
+    finished_product_id = fields.Many2one(
+        comodel_name="product.template",
+        domain="[('finished_product', '=', True)]",
+        required=True,
+    )
+    line_ids = fields.One2many(
+        comodel_name="finished.product.structure.line", inverse_name="structure_id"
+    )
+    attribute_ids = fields.Many2many(
+        comodel_name="product.attribute", compute="_compute_attribute_ids"
+    )
+    need_confirmation = fields.Boolean()
+
+    @api.onchange("finished_product_id")
+    def _onchange_finished_product_id(self):
+        """
+        Update attribute and line attributes
+        by onchange finished product
+        """
+        if self:
+            self._compute_attribute_ids()
+            self.line_ids.write({"attribute_ids": [(6, 0, self.attribute_ids.ids)]})
+            self.need_confirmation = bool(
+                self.finished_product_id.semi_finished_mrp_bom_ids
+            )
+
+    def _compute_attribute_ids(self):
+        for rec in self:
+            rec.attribute_ids = rec.finished_product_id.attribute_line_ids.mapped(
+                "attribute_id"
+            )
+
+    def _prepare_new_product_templates(self):
+        finished_product = self.finished_product_id
+        for line in self.line_ids:
+            new_product_name = "{} - {}".format(finished_product.name, line.stage_name)
+            product_tmpl = line.product_tmpl_id.copy({"name": new_product_name})
+            attr_lines = finished_product.attribute_line_ids.filtered(
+                lambda pl: pl.attribute_id in line.attribute_ids
+            )
+            for attr_line in attr_lines:
+                attr_line.copy({"product_tmpl_id": product_tmpl.id})
+            line.product_tmpl_stage_id = product_tmpl
+
+    def _tmp_product_struct_line(self):
+        self.ensure_one()
+        return self.env["finished.product.structure.line"].new(
+            {
+                "stage_name": "{} - Start".format(self.finished_product_id.name),
+                "product_tmpl_id": self.finished_product_id,
+                "product_tmpl_stage_id": self.finished_product_id,
+            }
+        )
+
+    def _prepare_bom_by_products(self):
+        lines = self._tmp_product_struct_line() | self.line_ids
+        vals_list = []
+        for index in range(len(lines)):
+            try:
+                bom = lines[index]
+                line = lines[index + 1]
```

Comment:
**suggestion (non-blocking)**: I think this bit of code could be clearer if you used itertools.pairwise https://docs.python.org/3/library/itertools.html#itertools.pairwise

And it would allow you to remove the try-except and the `range(len(lines))`

Reply:
Thanks, I fixed it. But in Python v3.7 doesn't have this function in `itertools` library. I implemented this function in code.